server/sonar-server/src/test/java/org/sonar/server/computation/step/QualityGateMeasuresStepTest.java

```diff
@@ -0,0 +1,263 @@
+    when(measureRepository.getRawMeasure(PROJECT_COMPONENT, INT_METRIC_1)).thenReturn(Optional.of(rawMeasure));
+
+    underTest.execute();
+
+    ArgumentCaptor<Measure> equals2ConditionMeasureCaptor = ArgumentCaptor.forClass(Measure.class);
+
+    verify(measureRepository).getRawMeasure(PROJECT_COMPONENT, INT_METRIC_1);
+    verify(measureRepository).update(same(PROJECT_COMPONENT), same(INT_METRIC_1), equals2ConditionMeasureCaptor.capture());
+    verify(measureRepository).add(same(PROJECT_COMPONENT), same(ALERT_STATUS_METRIC), alertStatusMeasureCaptor.capture());
+    verify(measureRepository).add(same(PROJECT_COMPONENT), same(QUALITY_GATE_DETAILS_METRIC), qgDetailsMeasureCaptor.capture());
+    verifyNoMoreInteractions(measureRepository);
+
+    assertThat(equals2ConditionMeasureCaptor.getValue())
+      .hasQualityGateLevel(OK)
+      .hasQualityGateText(dumbResultTextAnswer(equals2Condition, OK, rawValue));
+    assertThat(alertStatusMeasureCaptor.getValue())
+      .hasQualityGateLevel(OK)
+      .hasQualityGateText(dumbResultTextAnswer(equals2Condition, OK, rawValue));
+    assertThat(qgDetailsMeasureCaptor.getValue())
+      .hasValue(new QualityGateDetailsData(OK, of(new EvaluatedCondition(equals2Condition, OK, rawValue))).toJson());
+  }
+
+  @Test
+  public void new_measures_have_ERROR_level_if_at_least_one_updated_measure_has_ERROR_level() {
+    int rawValue = 1;
+    Condition equals1ErrorCondition = createEqualsCondition(INT_METRIC_1, "1", null);
+    Condition equals1WarningCondition = createEqualsCondition(INT_METRIC_2, null, "1");
+    Measure rawMeasure = Measure.newMeasureBuilder().create(rawValue, null);
+
+    qualityGateHolder.setQualityGate(new QualityGate("name", of(equals1ErrorCondition, equals1WarningCondition)));
+    when(measureRepository.getRawMeasure(PROJECT_COMPONENT, INT_METRIC_1)).thenReturn(Optional.of(rawMeasure));
+    when(measureRepository.getRawMeasure(PROJECT_COMPONENT, INT_METRIC_2)).thenReturn(Optional.of(rawMeasure));
+
+    underTest.execute();
+
+    ArgumentCaptor<Measure> equals1ErrorConditionMeasureCaptor = ArgumentCaptor.forClass(Measure.class);
+    ArgumentCaptor<Measure> equals1WarningConditionMeasureCaptor = ArgumentCaptor.forClass(Measure.class);
+
+    verify(measureRepository).getRawMeasure(PROJECT_COMPONENT, INT_METRIC_1);
+    verify(measureRepository).getRawMeasure(PROJECT_COMPONENT, INT_METRIC_2);
+    verify(measureRepository).update(same(PROJECT_COMPONENT), same(INT_METRIC_1), equals1ErrorConditionMeasureCaptor.capture());
+    verify(measureRepository).update(same(PROJECT_COMPONENT), same(INT_METRIC_2), equals1WarningConditionMeasureCaptor.capture());
+    verify(measureRepository).add(same(PROJECT_COMPONENT), same(ALERT_STATUS_METRIC), alertStatusMeasureCaptor.capture());
+    verify(measureRepository).add(same(PROJECT_COMPONENT), same(QUALITY_GATE_DETAILS_METRIC), qgDetailsMeasureCaptor.capture());
+    verifyNoMoreInteractions(measureRepository);
+
```

Comment:
as the `verifyNoMoreInteractions` highlights, I want to verify the implementation (I know, it's a tight unit test). It is not supposed to do any else than call these specific add, update and getRawMeasure methods. If anyone changes that, I want the test to break.